NMS-12892:  Document the MailTransportMonitor (#3143)
OpenNMS/opennms: opennms-doc/guide-admin/src/asciidoc/text/service-assurance/monitors/MailTransportMonitor.adoc

// Allow GitHub image rendering
:imagesdir: ../../../images

==== MailTransportMonitor

> MailTransportMonitor is used to run a synthetic test of a complete email transaction, including sending a mail and determining that it has been delivered.  It can also use both sendmail-test and readmail-test 

Reviewer: Text is great. I've been trying to standardize so that there is a separate line for each sentence.

independently to determine whether an email can be sent or a mailbox can be read.

===== Monitor facts

[options="autowidth"]
|===
| Class Name     | `org.opennms.netmgt.poller.monitors.MailTransportMonitor`
| Remote Enabled | true
|===

===== Configuration and Usage

.Monitor specific parameters for the MailTransportMonitor
[options="header, autowidth"]
|===
| Parameter        | Description                                                                                       | Required | Default value

> | `sendmail-test` | Defines the test for sending mail, contains sendmail-host, sendmail-protocol, sendmail-message,

Reviewer: ```suggestion
| `sendmail-test` | Defines the test for sending mail. Contains sendmail-host, sendmail-protocol, sendmail-message,
```

                     and user-auth                                                                                     | optional | `debug="true" use-authentication="false" use-jmta="true" attempt-interval="3000"`

> | `readmail-test` | Defines the test for reading mail, contains readmail-host, readmail-protocol, and user-auth.       | optional | `debug="true" mail-folder="INBOX" attempt-interval="1000" delete-all-mail="false"`

Reviewer: ```suggestion
| `readmail-test` | Defines the test for reading mail. Contains readmail-host, readmail-protocol, and user-auth.       | optional | `debug="true" mail-folder="INBOX" attempt-interval="1000" delete-all-mail="false"`
```

| `use-jmta`           | Use the JavaMail Mail Transport Agent                                                          | optional | `true`
| `attempt-interval`   | Interval in ms between send attempts                                                           | optional | `3000`
|===

.Options for sendmail-host
[options="header, autowidth"]
|===
| Attribute       | Description                                                                                         | Required | Default value
| `host`          | The SMTP server address for sending mail                                                            | optional | `${ipaddr}`
| `port`          | The SMTP server port                                                                                | optional | `25`
|===

.Options for sendmail-protocol
[options="header, autowidth"]
|===
| Attribute              | Description                                                                                  | Required | Default value
| `char-set`             | Set the character set                                                                        | optional | `us-ascii`

> | `mailer`               | use smtpsend or some alternate mailer                                                        | optional | `smtpsend`

Reviewer: ```suggestion
| `mailer`               | Use smtpsend or an alternate mailer                                                        | optional | `smtpsend`
```

| `message-content-type` | Set the message content-type                                                                 | optional | `text/plain`
| `message-encoding`     | Set the message encoding                                                                     | optional | `7-bit`
| `quit-wait`            | If set to false, the QUIT command is sent and the connection is immediately closed. If set 
                            to true (the default), causes the transport to wait for the response to the QUIT command.   | optional | `true`

> | `transport`            | smtp  or smtps                                                                               | optional | `smtp`

Reviewer: ```suggestion
| `transport`            | SMTP  or SMTPS                                                                               | optional | `smtp`
```

Author: These are literals.  Should that be `smtp` or `smtps` in that case?

Reviewer: lower case if they are literals.

| `debug`            | Show additional debug output                                                                     | optional | `true`
| `mail-folder`      | The folder or IMAP label to check for mail                                                       | optional | `INBOX`
| `subject-match`    | Substring match in email subjects when looking for a specific email                              | optional | `no default`
| `attempt-interval` | Attempt to read email after this many milliseconds have passed, also used for retry interval     | optional | `1000`
| `delete-all-mail`  | Delete all read mail after a successful match                                                    | optional | `false`
|===

.Options for readmail-host
[options="header, autowidth"]
|===
| Attribute   | Description                                                                                             | Required | Default value
| `host`      | The target host for reading mail                                                                        | optional | `${ipaddr}`

> | `port`      | the appropriate port for the protocol                                                                   | optional | `110`

Reviewer: ```suggestion
| `port`      | The appropriate port for the protocol                                                                   | optional | `110`
```

|===

.Options for readmail-protocol
[options="header, autowidth"]
|===
| Attribute    | Description                                                                                             | Required | Default value

> | `transport`  | The transport ptotocol to use.  One of:  pop3 pop3s imap imaps                                          | optional | `pop3`

Reviewer: ```suggestion
| `transport`  | The transport ptotocol to use. One of: pop3 pop3s imap imaps                                          | optional | `pop3`
```
Are these literals? If not, please capitalize: POP3 POP3S IMAP IMAPS

Author: Yes, these are literal.

          </readmail-host>
          <user-auth user-name="opennms@gmail.com" password="opennms"/>
        </readmail-test>
      </mail-test>
    </mail-transport-test>
  </parameter>
  <parameter key="rrd-repository" value="/opt/opennms/share/rrd/response"/>
  <parameter key="ds-name" value="mtm_lat"/>
  <parameter key="retry" value="20" />
</service>
----

Test that we can connect via IMAPS and open the OPENNMS folder.

[source, xml]
----
<service name="MTM-Readmail" interval="300000" user-defined="false" status="on">
  <parameter key="mail-transport-test">
    <mail-transport-test>
      <mail-test>
        <readmail-test attempt-interval="5000" mail-folder="OPENNMS" debug="false" >
          <readmail-host host="imap.gmail.com" port="993">
            <readmail-protocol ssl-enable="true" start-tls="false" transport="imaps" />
          </readmail-host>
          <user-auth user-name="opennms@gmail.com" password="opennms"/>
        </readmail-test>
      </mail-test>
    </mail-transport-test>
  </parameter>
  <parameter key="rrd-repository" value="/opt/opennms/share/rrd/response"/>
  <parameter key="ds-name" value="rdmail_lat"/>
</service>
----

===== Tests

> There are five basic tests that can be performed by this monitor.

Reviewer: ```suggestion
There are five basic tests that this monitor can perform.
```

Sending Mail: The most basic test, the sendmail-test is highly configurable. An Exception thrown during the sending of the configured email message will cause the poll to fail.

Access of Mail Store and Folder: Configure a readmail-test and don't configure a subject-match attribute. This will test only the ability to open the default mail store and the configured mail folder ("INBOX") by default. Folders are given by "INBOX<seperator>Foldername"; seperator character might vary between IMAP implementations. Exchange uses "/" as seperator, for example.

Specific Message in Folder: Configure a readmail-test and a matching subject. Optionally configure the test to delete all read mail. Probably not something you want to do unless this is a mail folder that you are sending email to from another system that you can't do in the end-to-end test behavior.

> Sending and Receipt (end-to-end test) of a Message:Test your infrastructure's ability to send and receive email. It tests sending and receiving of an mail message via one or two separate mail servers. For example, you can send email via SMTPS to one server outside of your organization addressed to a recipient on your internal mail server and verify delivery.

Reviewer: ```suggestion
Sending and Receipt (end-to-end test) of a Message: Test your infrastructure's ability to send and receive email. It tests sending and receiving of an mail message via one or two separate mail servers. For example, you can send email via SMTPS to one server outside of your organization addressed to a recipient on your internal mail server and verify delivery.
```
Separate lines, if possible.

.Options for readmail-protocol
[options="header, autowidth"]
|===
| Attribute    | Description                                                                                            | Required | Default value

> | `transport`  | The transport ptotocol to use. One of: pop3 pop3s imap imaps                                           | optional | `pop3`

Reviewer: ```suggestion
| `transport`  | The transport protocol to use. One of: pop3 pop3s imap imaps                                           | optional | `pop3`
```

> | `transport`  | The transport ptotocol to use. One of: pop3 pop3s imap imaps                                           | optional | `pop3`

Reviewer: ```suggestion
| `transport`  | The transport protocol to use. One of: pop3 pop3s imap imaps                                           | optional | `pop3`
```

----

===== Tests
There are five basic tests that this monitor can perform.

Sending Mail: The most basic test, the sendmail-test is highly configurable.

> An Exception thrown during the sending of the configured email message will cause the poll to fail.

Reviewer: ```suggestion
An exception thrown during the sending of the configured email message will cause the poll to fail.
```
